hplsql/src/main/java/org/apache/hive/hplsql/packages/HmsPackageRegistry.java

```
@@ -0,0 +1,103 @@
+
+import java.util.Optional;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.AddPackageRequest;
+import org.apache.hadoop.hive.metastore.api.DropPackageRequest;
+import org.apache.hadoop.hive.metastore.api.GetPackageRequest;
+import org.apache.hadoop.hive.metastore.api.Package;
+import org.apache.hive.hplsql.HplSqlSessionState;
+import org.apache.thrift.TException;
+
+public class HmsPackageRegistry implements PackageRegistry {
+  private final IMetaStoreClient msc;
+  private final HplSqlSessionState hplSqlSession;
+
+  public HmsPackageRegistry(IMetaStoreClient msc, HplSqlSessionState hplSqlSession) {
+    this.msc = msc;
+    this.hplSqlSession = hplSqlSession;
+  }
+
+  @Override
+  public Optional<String> getPackage(String name) {
+    try {
+      Package pkg = msc.findPackage(request(name));
+      return pkg == null
+              ? Optional.empty()
+              : Optional.of(pkg.getHeader() + ";\n" + pkg.getBody());
+    } catch (TException e) {
+      throw new RuntimeException(e.getCause());
+    }
+  }
+
+  @Override
+  public void createPackageHeader(String name, String header, boolean replace) {
+    try {
+      Package existing = msc.findPackage(request(name));
+      if (existing != null && !replace)
+        throw new RuntimeException("Package " + name + " already exists");
+      msc.addPackage(makePackage(name, header, ""));
+    } catch (TException e) {
+      throw new RuntimeException(e.getCause());
+    }
+  }
+
+  @Override
+  public void createPackageBody(String name, String body, boolean replace) {
+    try {
+      Package existing = msc.findPackage(request(name));
+      if (existing != null && StringUtils.isNotEmpty(existing.getBody()) && !replace)
+        throw new RuntimeException("Package body " + name + " already exists");
+      if (existing == null) {
+        msc.addPackage(makePackage(name, "", body));
```

Is a package body without header valid? I believe this should throw an error.
https://docs.oracle.com/cd/E11882_01/appdev.112/e25519/packages.htm#LNPLS00901 says "A package always has a specification"

---

HPLSQL used to allow this in previous versions. Oracle also allows you to create the body without the spec but the package is not usable until you create the spec for it. Throwing an error at this point wouldn't be the same either. We might want to put a check when getting the package, but it would break the previous behaviour. I suspect the limitation in oracle is coming from the fact that it compiles these two together but since this language is fully interpreted we don't face this problem. An other option would be to print out a warning message. Or just leave it as it is since this is how it used to work. Which one do you think is better?

---

But our design does not allow creating a body than creating a header for it. `createPackageHeader` queries HMS with the package name. If it finds the package, it always overrides the body.

---

Ok, let's go with not allowing headerless packages.